FINAL

iskachkov opened the pull request.

iskachkov committed:
Project DNA

iskachkov committed:
Теперь нуклеотид всегда заменяется на другой

iskachkov committed:
Проверка входных параметров

skorochkin commented:
Несколько моих замечаний:
- нецелесообразность выноса **nucleotides** в проперти + добавть метод который будет возвращать случайный элемент из этого **nucleotides**
- на сколько я понял objective-c предполагает обращение к элементам массива с помощью, например characterAtIndex или replaceElementAtIndex вместо array[index]
- не совсем корректное округление таким образом. (int)3.9 = 3
- цикл можно уменьшить до (percent / 100 \* lenght) итераций. Нет смысла каждый раз просматривать весь массив
- (мелкое) массив объявлять в одну строку: blabla *b = [[blabla alloc] init];
- (мелкое) проверка процент (0 > percent < 100) это хорошою, защита от идиота, но я не уверен в ее целесообразности сейчас, когда процент указывается явно в коде

iskachkov committed:
Рефакторинг

iskachkov committed:
Корректное округление до целого числа

iskachkov commented:
> нецелесообразность выноса nucleotides в проперти + добавть метод который будет возвращать случайный элемент из этого nucleotides

Исправлено! Действительно, кода стало меньше.

> на сколько я понял objective-c предполагает обращение к элементам массива с помощью, например characterAtIndex или replaceElementAtIndex вместо array[index]

Похоже на то. Исправил. Но, на мой взгляд, код стало сложнее читать. Может еще просто не привык. 

> не совсем корректное округление таким образом. (int)3.9 = 3

Исправлено!

> цикл можно уменьшить до (percent / 100 \* lenght) итераций. Нет смысла каждый раз просматривать весь массив

Вот этот момент не понял. В чем разница с (length / 100 \* percent)? 

> (мелкое) массив объявлять в одну строку: blabla *b = [[blabla alloc] init];

Исправлено!

> (мелкое) проверка процент (0 > percent < 100) это хорошою, защита от идиота, но я не уверен в ее целесообразности сейчас, когда процент указывается явно в коде

Программист тоже может быть идиотом :-) Да и от опечаток никто не застрахован. Я за проверку входящих значений в любом случае.

skorochkin commented:
Сложнее читать - не сказал бы, т.к. если ты помнишь Рахим на лекции говорил, что код objective-c сделан так, чтобы читать его как фразы на английском языке.

По поводу цикла - по всей видимости я ошибся (мне казалось что в **mutate** у тебя был цикл по всему массиву DNA), так что не обращай внимания! :)

iskachkov commented:
> Сложнее читать - не сказал бы, т.к. если ты помнишь Рахим на лекции говорил, что код objective-c сделан так, чтобы читать его как фразы на английском языке.

Ну да. Скорее всего просто не привык :-)

iskachkov committed:
FINAL

Dante777 commented:
Несколько замечаний:
- Массив DNA получается не 100 символов, а 101 - по коду цикл заполнения от 0 до 100 включительно.
- Нет вывода цепочек в консоль - сложно проверять
- В mutate для вычисления числа замен используется целочисленная арифметика, в результате при длине цепи, отличной от 100 (например, 4 - для проверки), count == 0 при любом проценте, и цикл отработает только один раз

iskachkov commented:
Спасибо за замечания. Особенно про длину массива. Серьезная ошибка.  
У себя все исправил. Не знаю насколько корректно будет сюда заливать изменения. Ведь уже поставил слово FINAL.

Dante777 commented:
Мне работу прислали на проверку буквально час назад, думаю, преподаватели это примут во внимание. Можно, конечно, уточнить в дискуссии по работе.

iskachkov committed:
FINAL bug fixes

Если принципиально оценить именно FINAL commit, то оценивать нужно
предыдущий.

- Исправлена ошибка с кол-вом элементов ДНК
- Добавлен метод print для Cell
- Для рассчета кол-ва изменяемых элементов в mutate теперь используется
float вместо int

iskachkov commented:
Залил новую версию с пометкой о баг фиксе финальной версии.

Dante777 commented:
Замени еще в методе мутации i <= count на i < count, а то получается лишний прогон, и все будет Ок.
Есть еще мелкие замечания по стилю, но это к заданию не относится )

iskachkov committed:
FINAL bug fixes

iskachkov commented:
Исправил.

По стилю тоже было бы интересно послушать замечания :-)

Dante777 commented:
Ну так, навскидку :)
- константы типа длины массива обычно определяются в .h файлах, а не в коде
- описание и реализацию категории лучше выносить в отдельные файлы, а не в main
- когда функция бросает Exception, его надо ловить и обрабатывать.

Dante777 commented:
10

slebedev commented on DNA/DNA/main.m:
```diff
@@ -0,0 +1,63 @@
+-(void) mutate:(int)percent {
+    // Проверяем входящие параметры
+    if (percent > 100 || percent < 1) {
+        [NSException raise:@"Invalid percent value" format:@"percent of %d is invalid", percent];
```
Меня вот проверяющий пнул, что я 0 отбрасываю. =) По идее, ничего страшного в нуле нет, цикл не должен будет отрабатывать, а соответственно, ничего не должно поменяться.

iskachkov committed:
...

- Изменен deployment_target на 10.6
- Категория mutator вынесена в отдельные файлы
- Длина ДНК задается теперь константой
- Мутация на 0 процентов теперь не является ошибкой
